csharp/src/Apache.Arrow.Adbc/Tracing/ActivityTrace.cs

```diff
@@ -0,0 +1,325 @@
+        /// Activity status is set to <see cref="ActivityStatusCode.Ok"/> and the result is returned.
+        /// If an exception is thrown by the delegate, the Activity status is set to <see cref="ActivityStatusCode.Error"/>
+        /// and an Event <see cref="ActivityEvent"/> is added to the actitity and finally the exception is rethrown.
+        /// </remarks>
+        public static async Task<T> TraceActivityAsync<T>(ActivitySource activitySource, Func<Activity?, Task<T>> call, [CallerMemberName] string? activityName = default, string? traceParent = default)
+        {
+            using Activity? activity = StartActivityInternal(activityName, activitySource, traceParent);
+            try
+            {
+                T? result = await call(activity);
+                if (activity?.Status == ActivityStatusCode.Unset) activity?.SetStatus(ActivityStatusCode.Ok);
+                return result;
+            }
+            catch (Exception ex)
+            {
+                TraceException(ex, activity);
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the trace parent context.
+        /// </summary>
+        public string? TraceParent { get; set; }
+
+        /// <summary>
+        /// Writes the exception to the trace by adding an exception event to the current activity (span).
+        /// </summary>
+        /// <param name="exception">The exception to trace.</param>
+        /// <param name="activity">The current activity where the exception is caught.</param>
+        /// <param name="escaped">
+        /// An indicator that should be set to true if the exception event is recorded
+        /// at a point where it is known that the exception is escaping the scope of the span/activity.
+        /// For example, <c>escaped</c> should be <c>true</c> if the exception is caught and re-thrown.
+        /// However, <c>escaped</c> should be set to <c>false</c> if execution continues in the current scope.
+        /// </param>
+        private static void TraceException(Exception exception, Activity? activity) =>
+            WriteTraceException(exception, activity);
+
+        /// <summary>
+        /// Disposes managed and unmanaged objects. If overridden, ensure to call this base method.
+        /// </summary>
+        /// <param name="disposing">An indicator of whether this method is being called from the <c>Dispose</c> method.</param>
+        private void Dispose(bool disposing)
+        {
+            if (!_isDisposed && disposing)
+            {
+                ActivitySource.Dispose();
+                _isDisposed = true;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            // Do not change this code. Put cleanup code in 'Dispose(bool disposing)' method
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
```

Comment:
There's no reason to call `GC.SuppressFinalize` on an instance of a class that doesn't implement a finalizer. Also applies to `TracingReader` and `TracingConnection`. And I think in general, there isn't really a reason to implement the traditional Dispose pattern unless the class is holding an unmanaged resource and has a finalizer.